Cubical/Data/FinSet/Properties.agda

```diff
@@ -0,0 +1,253 @@
+
+  isFinSetΣ : isFinSet (Σ (X .fst) (λ x → Y x .fst))
+  isFinSetΣ =
+    elim2 (λ _ _ → isProp-isFinSet {A = Σ (X .fst) (λ x → Y x .fst)})
+          (λ p q → ∣ ≃FinΣ (X .fst) p (λ x → Y x .fst) q ∣)
+          (X .snd) (choice X (λ x → ≃Fin (Y x .fst)) (λ x → Y x .snd))
+
+  isFinSetΠ : isFinSet ((x : X .fst) → Y x .fst)
+  isFinSetΠ =
+    elim2 (λ _ _ → isProp-isFinSet {A = ((x : X .fst) → Y x .fst)})
+          (λ p q → ∣ ≃FinΠ (X .fst) p (λ x → Y x .fst) q ∣)
+          (X .snd) (choice X (λ x → ≃Fin (Y x .fst)) (λ x → Y x .snd))
+
+module _
+  (X : FinSet {ℓ})
+  (Y : X .fst → FinSet {ℓ'})
+  (Z : (x : X .fst) → Y x .fst → FinSet {ℓ''}) where
+
+  isFinSetΠ2 : isFinSet ((x : X .fst) → (y : Y x .fst) → Z x y .fst)
+  isFinSetΠ2 = isFinSetΠ X (λ x → _ , isFinSetΠ (Y x) (Z x))
+
+module _
+  (X : FinSet {ℓ})
+  (Y : X .fst → FinSet {ℓ'})
+  (Z : (x : X .fst) → Y x .fst → FinSet {ℓ''})
+  (W : (x : X .fst) → (y : Y x .fst) → Z x y .fst → FinSet {ℓ'''}) where
+
+  isFinSetΠ3 : isFinSet ((x : X .fst) → (y : Y x .fst) → (z : Z x y .fst) → W x y z .fst)
+  isFinSetΠ3 = isFinSetΠ X (λ x → _ , isFinSetΠ2 (Y x) (Z x) (W x))
+
+module _
+  (X : FinSet {ℓ}) where
+
+  isFinSet≡ : (a b : X .fst) → isFinSet (a ≡ b)
+  isFinSet≡ a b = isDecProp→isFinSet (isFinSet→isSet (X .snd) a b) (isFinSet→Discrete (X .snd) a b)
+
+  isFinSetIsContr : isFinSet (isContr (X .fst))
+  isFinSetIsContr = isFinSetΣ X (λ x → _ , (isFinSetΠ X (λ y → _ , isFinSet≡ x y)))
+
+  isFinSet∥∥ : isFinSet ∥ X .fst ∥
+  isFinSet∥∥ = TruncRec isProp-isFinSet (λ p → ∣ ≃Fin∥∥ (X .fst) p ∣) (X .snd)
+
+module _
+  (X : FinSet {ℓ })
+  (Y : FinSet {ℓ'})
+  (f : X .fst → Y .fst) where
+
+  isFinSetFiber : (y : Y .fst) → isFinSet (fiber f y)
+  isFinSetFiber y = isFinSetΣ X (λ x → _ , isFinSet≡ Y (f x) y)
+
+  isFinSetIsEquiv : isFinSet (isEquiv f)
+  isFinSetIsEquiv =
+    EquivPresFinSet
+      (invEquiv (isEquiv≃isEquiv' f))
+      (isFinSetΠ Y (λ y → _ , isFinSetIsContr (_ , isFinSetFiber y)))
+
+module _
+  (X : FinSet {ℓ })
+  (Y : FinSet {ℓ'}) where
+
+  isFinSet⊎ : isFinSet (X .fst ⊎ Y .fst)
+  isFinSet⊎ = elim2 (λ _ _ → isProp-isFinSet) (λ p q → ∣ ≃Fin⊎ (X .fst) p (Y .fst) q ∣) (X .snd) (Y .snd)
+
+  isFinSet× : isFinSet (X .fst × Y .fst)
+  isFinSet× = elim2 (λ _ _ → isProp-isFinSet) (λ p q → ∣ ≃Fin× (X .fst) p (Y .fst) q ∣) (X .snd) (Y .snd)
+
+  isFinSet→ : isFinSet (X .fst → Y .fst)
+  isFinSet→ = isFinSetΠ X (λ _ → Y)
+
+  isFinSet≃ : isFinSet (X .fst ≃ Y .fst)
+  isFinSet≃ = isFinSetΣ (_ , isFinSet→) (λ f → _ , isFinSetIsEquiv X Y f)
+
+module _
+  (X : FinSet {ℓ}) where
+
+  isFinSet¬ : isFinSet (¬ (X .fst))
+  isFinSet¬ = isFinSet→ X (⊥ , ∣ 0 , uninhabEquiv (λ x → x) ¬Fin0 ∣)
+
+module _
+  (X : FinSet {ℓ}) where
+
+  isFinSetNonEmpty : isFinSet (NonEmpty (X .fst))
+  isFinSetNonEmpty = isFinSet¬ (_ , isFinSet¬ X)
+
+module _
+  {X : Type ℓ }
+  {Y : Type ℓ'}
+  (f : X → Y) where
+
+  isInjective : Type (ℓ-max ℓ ℓ')
+  isInjective = (a b : X) → ¬ (a ≡ b) → ¬ (f a ≡ f b)
```

Comment:
This is a roundabout way to define injectivity? We should probably upstream this to a suitable place as well

Reply:
A suitable place would be in `Cubical.Functions`

Reply:
Ad-hoc but direct. I think it's what you can find in any basic textbooks but is only working for Sets. Maybe I should turn to those definitions in `Cubical.Functions`.

Reply:
Right, but a much nicer version for constructive math is to express it positively as: `f a ≡ f b → a ≡ b`

There are many uses of this in the library, but not expressed as `isInjective f`...